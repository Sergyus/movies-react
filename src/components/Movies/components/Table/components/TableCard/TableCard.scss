@import "../../../../../../styles/variables";

.card {
  width: 100%;
  box-sizing: border-box;
  padding: 30px;
  margin-bottom: 35px;
  background-color: #fff;
  position: relative;
  display: flex;
  box-shadow: 0 10px 40px 0 rgba(0, 0, 0, 0.3);

  & img {
    flex-grow: 1;
    height: 320px;
    max-width: 240px;
    object-fit: cover;
    margin-right: 30px;
  }
}

.info {
  text-align: left;
  flex-grow: 10;

  & > span {
    font-family: "Helvetica Neue", Arial, Helvetica, sans-serif;
    font-weight: bold;
    color: $blue-color;
    line-height: 1.4rem;
  }
}

.title {
  width: 100%;
  color: #242b2d;
  font-size: 1.4rem;
  text-align: center;
  padding-top: 10px;
  margin-bottom: 20px;

  & ~ span {
    display: inline-block;
    font-size: 1rem;
    margin-right: 20px;
  }
}

.overview {
  margin: 15px 0;
  line-height: 1.5rem;
}

.rating {
  margin-bottom: 25px;
  display: flex;

  & span {
    margin-left: 15px;
    font-size: 0.9rem;
    color: $blue-color;
  }
}

@media (max-width: $tablet-screen-size) {
  .card {
    flex-wrap: wrap;

    & img {
      margin: 0 auto;
      height: 300px;
      width: 100%;
    }
  }

  .info {
    text-align: center;
  }

  .title {
    font-size: 1.2rem;
    margin-bottom: 10px;

    & ~ span {
      font-size: 0.9rem;
    }
  }

  .overview {
    font-size: 0.9rem;
  }

  .rating {
    display: flex;
    justify-content: center;

    & span {
      font-size: 0.8rem;
    }
  }
}
