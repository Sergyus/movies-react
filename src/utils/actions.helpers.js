export const createAsyncFnRequest = requestName => ({
  type: requestName,
});

export const createAsyncFnFailure = (requestName, error) => ({
  type: requestName,
  error: error.message,
});
