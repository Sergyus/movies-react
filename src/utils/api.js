const MOVIE_DB_API_KEY = "ed98d25919802b5020e8feb6c3923886";
const MOVIE_DB_BASE_URL = "https://api.themoviedb.org/3";
const LANGUAGE = "ru-RU";

const createMovieDbUrl = (relativeUrl, queryParams) => {
  let baseUrl = `${MOVIE_DB_BASE_URL}${relativeUrl}?api_key=${MOVIE_DB_API_KEY}&language=${LANGUAGE}`;
  if (queryParams) {
    Object.keys(queryParams).forEach(paramName => {
      baseUrl += `&${paramName}=${queryParams[paramName]}`;
      return baseUrl;
    });
  }
  return baseUrl;
};

const getLatestMovies = ({ page }) => {
  const fullUrl = createMovieDbUrl("/movie/popular", {
    page
  });
  return fetch(fullUrl);
};

const getTopMovies = ({ page }) => {
  const fullUrl = createMovieDbUrl("/movie/top_rated", {
    page
  });
  return fetch(fullUrl);
};

const getUpcomingMovies = ({ page }) => {
  const fullUrl = createMovieDbUrl("/movie/upcoming", {
    page
  });
  return fetch(fullUrl);
};

const searchMovies = ({ page, query }) => {
  const fullUrl = `${createMovieDbUrl("/search/movie", {
    page,
    query
  })}&include_adult=false`;
  return fetch(fullUrl);
};

const getGenres = () => {
  const fullUrl = createMovieDbUrl("/genre/movie/list");
  return fetch(fullUrl);
};

const getMoviesByGenre = ({ page, genre }) => {
  const fullUrl = `${createMovieDbUrl("/discover/movie", {
    page
  })}&sort_by=popularity.desc&include_adult=false&include_video=false&with_genres=${genre}`;
  return fetch(fullUrl);
};

const getMovieDetails = ({ movieId }) => {
  const fullUrl = createMovieDbUrl(`/movie/${movieId}`);
  return fetch(fullUrl);
};

const getMovieVideo = movieId => {
  const fullUrl = createMovieDbUrl(`/movie/${movieId}/videos`);
  return fetch(fullUrl);
};

export {
  getTopMovies,
  searchMovies,
  getMovieDetails,
  getGenres,
  getMoviesByGenre,
  getLatestMovies,
  getUpcomingMovies,
  getMovieVideo
};
