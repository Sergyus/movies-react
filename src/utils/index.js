import {
  getTopMovies, searchMovies, getGenres, getMoviesByGenre,
  getLatestMovies, getUpcomingMovies, getMovieVideo,
} from './api';
import { getMovies } from './tbdbApiService';
import truncate from './truncate';
import { useFetchOnScroll } from './useFetchOnScroll';
import useBodyScrollLock from './useBodyScrollLock';
import useOnClickOutside from './useOnClickOutside';
import { createAsyncFnFailure, createAsyncFnRequest } from './actions.helpers';

export {
  getTopMovies, searchMovies, getGenres, getMoviesByGenre,
  getLatestMovies, getUpcomingMovies, getMovieVideo,
  truncate, useBodyScrollLock, useOnClickOutside, getMovies,
  createAsyncFnFailure, createAsyncFnRequest, useFetchOnScroll,
};
