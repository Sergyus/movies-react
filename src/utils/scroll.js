const getScrollDownPercentage = ({ scrollHeight, scrollPos, clientHeight }) => {
  const currentPosition = scrollPos + clientHeight;
  const percentageScrolled = currentPosition / scrollHeight;
  return percentageScrolled;
};

export default getScrollDownPercentage;
