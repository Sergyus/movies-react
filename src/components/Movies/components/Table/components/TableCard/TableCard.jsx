import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import styles from './TableCard.scss';
import NumRating from '../../../../../NumRating';
import Genres from '../../../../../Genres';
import Button from '../../../../../Button';
import StarRating from '../../../../../StarRating';
import { truncate } from '../../../../../../utils';

const TableCard = ({ movie, fetchVideoSrc, toggleModal }) => {
  const {
    id,
    rating,
    title,
    genres,
    overview,
    image,
  } = movie;
  const maxLength = 500;
  const handleClick = useCallback(() => {
    fetchVideoSrc(id);
    toggleModal(true);
  }, [toggleModal, id, fetchVideoSrc]);
  return (
    <article className={styles.card}>
      <img src={image} alt="Poster" />
      <div className={styles.info}>
        <h2 className={styles.title}>{title}</h2>
        <Genres genres={genres} />
        <p className={styles.overview}>{truncate(overview, maxLength)}</p>
        <div className={styles.rating}>
          <StarRating starsSelected={rating} />
          <NumRating rating={rating} />
        </div>
        <Button type="primary" handleClick={handleClick}>Watch Now</Button>
      </div>
    </article>
  );
};

TableCard.propTypes = {
  movie: PropTypes.shape({
    rating: PropTypes.number,
    title: PropTypes.string,
    genres: PropTypes.array,
    overview: PropTypes.string,
    image: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  }).isRequired,
  toggleModal: PropTypes.func,
  fetchVideoSrc: PropTypes.func,
};

TableCard.defaultProps = {
  toggleModal: () => {},
  fetchVideoSrc: () => {},
};

export default TableCard;
