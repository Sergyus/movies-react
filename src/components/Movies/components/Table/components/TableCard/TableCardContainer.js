import { connect } from 'react-redux';
import TableCard from './TableCard';
import { fetchVideoSrc } from '../../../../../../modules/app';

export default connect(null, { fetchVideoSrc })(TableCard);
