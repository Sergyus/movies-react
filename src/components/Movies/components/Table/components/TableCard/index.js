import TableCardContainer from './TableCardContainer';

export default TableCardContainer;
