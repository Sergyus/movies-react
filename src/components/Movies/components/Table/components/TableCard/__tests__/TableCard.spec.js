import React from 'react';
import { create, act } from 'react-test-renderer';
import TableCard from '../TableCard';
import { getMovies } from '../../../../../../../utils';
import mockMovies from '../../../../../../../assets/json/popular.json';
import { genres } from '../../../../../../../assets/json/genres.json';

describe('TableCard component renders correctly', () => {
  const movie = { ...getMovies(mockMovies.results, genres)[0], id: 1 };

  it('TableCard renders correctly', () => {
    const tree = create(
      <TableCard
        movie={movie}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('TableCard watchBtnClick works correctly', () => {
    const tree = create(
      <TableCard
        movie={movie}
      />,
    );
    const watchBtn = tree.root.findByProps({ 'aria-label': 'Watch Now' });
    const mockWatchBtn = jest.fn(watchBtn.props.onClick);
    act(() => {
      mockWatchBtn();
    });
    expect(mockWatchBtn).toHaveBeenCalled();
  });
});
