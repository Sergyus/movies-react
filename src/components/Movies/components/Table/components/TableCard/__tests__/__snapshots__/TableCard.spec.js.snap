// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableCard component renders correctly TableCard renders correctly 1`] = `
<article
  className="card"
>
  <img
    alt="Poster"
    src="https://image.tmdb.org/t/p/w300/AnLCqcH66Ecu8VFmXu3RofK4SMK.jpg"
  />
  <div
    className="info"
  >
    <h2
      className="title"
    >
      T-34
    </h2>
    <span
      className="genre"
    >
      War
    </span>
    <span
      className="genre"
    >
      Drama
    </span>
    <span
      className="genre"
    >
      Adventure
    </span>
    <span
      className="genre"
    >
      Action
    </span>
    <p
      className="overview"
    >
      1941 - WWii. the second lieutenant Nikolai ivushkin, commander of a t-34, engages in an unequal battle against the tank ace Klaus Jager in a battle near moscow. His mission is more of a suicide - to destroy a dozen german tanks, all by himself. that said, luck does favour the bold. He wins the battle, barely survives, but loses his tank and lands himself in captivity for three long years... there was little to no chance for ivushkin and Jager to meet again, but the war knows how to throw a curve...
    </p>
    <div
      className="rating"
    >
      <div
        className="star"
      />
      <div
        className="star"
      />
      <div
        className="star"
      />
      <div
        className="star"
      />
      <div
        className="star"
      />
      <span
        className="numRating"
      >
        4.5
      </span>
    </div>
    <button
      aria-label="Watch Now"
      className="btn primary"
      onClick={[Function]}
      type="button"
    >
      Watch Now
    </button>
  </div>
</article>
`;
